README.md
# express-api-helper

Simple API helper module for [Express](http://expressjs.com) apps.

[![Build Status](https://secure.travis-ci.org/paambaati/express-api-helper.png)](http://travis-ci.org/paambaati/express-api-helper)

## API

### `ok(req, res, data)`

Respond with `200 OK` and JSON-encoded data.

* `req` express Request
* `res` express Response
* `data` Object

### `badRequest(req, res, errors)`

Respond with `400 Bad Request` and JSON-encoded error object, `{message:String,errors:Array}`.

* `req` express Request
* `res` express Response
* `data` Array (of String) or String

### `unauthorized(req, res)`

Respond with `401 Unauthorized` and JSON-encoded error object, `{message:String}`.

* `req` express Request
* `res` express Response

### `forbidden(req, res)`

Respond with `403 Forbidden` and JSON-encoded error object, `{message:String}`.

* `req` express Request
* `res` express Response

### `notFound(req, res)`

Respond with `404 Not Found` and JSON-encoded error object, `{message:String}`.

* `req` express Request
* `res` express Response

### `unsupportedAction(req, res)`

Respond with `405 Method Not Allowed` and JSON-encoded error object, `{message:String}`.

* `req` express Request
* `res` express Response

### `invalid(req, res, errors)`

Respond with `422 Unprocessable Entity` and JSON-encoded error object, `{message:String,errors:Array}`.

* `req` express Request
* `res` express Response
* `errors` Array (of String) or String

### `serverError(req, res, error)`

Respond with `500 Internal Server Error` and JSON-encoded error object, `{message:String,error:Object}`.

* `req` express Request
* `res` express Response
* `error` Object

### `requireParams(req, res, params, callback)`

Require that listed parameters are present. Checks for presence of each parameter in `req.body` object if using `express.bodyParser` middleware; otherwise checks for presence of each parameter in `req.params` or `req.query`. If any parameters are missing, invokes `badRequest` with an array of error messages with the form `"Missing required parameter: %s"`.

* `req` express Request
* `res` express Response
* `params` Array (of String) or String
* `callback(err)` Function

### `requireHeaders(req, res, headers, callback)`

Require that listed headers are present. Checks for presence of each header in `req.headers`. If any parameters are missing, invokes `badRequest` with an array of error messages with the form `"Missing required header parameter: %s"`.

* `req` express Request
* `res` express Response
* `headers` Array (of String) of String
* `callback(err)` Function 

## Example

Sample usage:

```javascript
var http = require('http'),
    express = require('express'),
    bodyParser = require('body-parser'),
    api = require('express-api-helper'),
    app = express(),
    Post = require('./models/post');

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({
    extended: true
}));

app.all('/api/*', function (req, res, next) {
  if (!req.user) return api.unauthorized(req, res);
  next();
});

app.post('/api/posts', function (req, res) {
  api.requireParams(req, res, ['title', 'content', 'authorId'], function (err) {
    if (err) return api.serverError(req, res, err);
    var payload = {
      title: req.body.title,
      content: req.body.content,
      authorId: req.body.authorId
    };
    Post.create(payload, function (err, post) {
      if (err) return api.serverError(req, res, err);
      api.ok(req, res, post.toJSON());
    });
  });
});

app.get('/api/posts', function (req, res) {
  Post.find({}, function (err, posts) {
    if (err) return api.serverError(req, res, err);
    api.ok(req, res, posts.toJSON());
  });
});

app.get('/api/posts/:id', function (req, res) {
  Post.findById(req.params.id, function (err, post) {
    if (err) return api.serverError(req, res, err);
    if (!post) return api.notFound(req, res);
    api.ok(req, res, post.toJSON());
  })
});

http.createServer(app).listen(3000, function () {
  console.log("Express API listening on 3000");
});
```

## Running Tests

To run the tests, clone the repository and install the dev dependencies:

```bash
git clone git://github.com/paambaati/express-api-helper.git
cd express-api-helper && npm install
make test
```

## License

[MIT](LICENSE)



CHANGELOG.md
# Change Log

## 0.0.5 (2015-07-17)

 * Add support for `403` Fobidden response.
 * Update testcases with Express 4.x changes (body-parser).
 * Bump dependencies - express 3.4.8 → 4.13.1, mocha 1.17.1 → 2.2.5, supertest 0.9.0 → 1.0.1.
 * Add Travis builds for Node 0.12 and iojs.

## 0.0.4 (2014-12-24)

 * Fix deprecation warnings thrown by Express 4.
 * Add requireHeaders() method to support checking for headers.

## 0.0.3 (2014-03-08)

  * When testing for required params in `requireParams()`, use `hasOwnProperty()` and check for params in `req.params` in addition to `req.body` and `req.query`

## 0.0.2 (2014-03-06)

  * Fix bug in `requireParams()` (Issue #1)
  * Upgrade dependency versions

## 0.0.1 (2013-03-10)

  * First version


test/api.test.js
'use strict';

var request = require('supertest'),
    expect = require('expect.js'),
    express = require('express'),
    bodyParser = require('body-parser'),
    api = require('../');

process.env.NODE_ENV = 'test';

describe('the api', function() {
    it('should respond with 200 status for valid requests', function(done) {
        var app = express(),
            data = {
                message: 'ok'
            };
        app.use(bodyParser.json());
        app.use(bodyParser.urlencoded({
            extended: true
        }));
        app.post('/api/test/:username', function(req, res) {
            api.requireParams(req, res, ['username', 'password', 'token'], function(err) {
                api.ok(req, res, {
                    message: 'ok'
                });
            });
        });

        request(app)
            .post('/api/test/hercules?token=0')
            .send({
                password: ''
            })
            .expect('Content-Type', /json/)
            .expect(200)
            .expect(JSON.stringify(data), done);
    });

    it('should respond with 400 status when POST request data is malformed', function(done) {
        var app = express(),
            expected = {
                message: 'Bad Request',
                errors: ['Missing required parameter: password']
            };
        app.use(bodyParser.json());
        app.use(bodyParser.urlencoded({
            extended: true
        }));
        app.post('/api/test', function(req, res) {
            api.requireParams(req, res, ['username', 'password'], function(err) {
                api.ok(req, res, {
                    message: 'ok'
                });
            });
        });

        request(app)
            .post('/api/test')
            .send({
                username: 'hercules'
            })
            .expect('Content-Type', /json/)
            .expect(400)
            .expect(JSON.stringify(expected), done);
    });

    it('should respond with 400 status when GET request data is malformed', function(done) {
        var app = express(),
            expected = {
                message: 'Bad Request',
                errors: ['Missing required parameter: password']
            };

        app.get('/api/test', function(req, res) {
            api.requireParams(req, res, ['username', 'password'], function(err) {
                api.ok(req, res, {
                    message: 'ok'
                });
            });
        });

        request(app)
            .get('/api/test?username=hercules')
            .expect('Content-Type', /json/)
            .expect(400)
            .expect(JSON.stringify(expected), done);
    });

    it('should respond with 400 status when required request headers are missing', function(done) {
        var app = express(),
            expected = {
                message: 'Bad Request',
                errors: ['Missing required header parameter: X-Auth-Token']
            };

        app.get('/api/test', function(req, res) {
            api.requireHeaders(req, res, ['X-Auth-Token'], function(err) {
                api.ok(req, res, {
                    message: 'ok'
                });
            });
        });

        request(app)
            .get('/api/test')
            .expect('Content-Type', /json/)
            .expect(400)
            .expect(JSON.stringify(expected), done);
    });

    it('should respond with 401 status when the client is unauthorized', function(done) {
        var app = express(),
            expected = {
                message: 'Unauthorized'
            };

        app.get('/api/test', function(req, res) {
            api.unauthorized(req, res);
        });

        request(app)
            .get('/api/test')
            .expect('Content-Type', /json/)
            .expect(401)
            .expect(JSON.stringify(expected), done);
    });

    it('should respond with 403 status when the client is forbidden', function(done) {
        var app = express(),
            expected = {
                message: 'Forbidden'
            };

        app.get('/api/test', function(req, res) {
            api.forbidden(req, res);
        });

        request(app)
            .get('/api/test')
            .expect('Content-Type', /json/)
            .expect(403)
            .expect(JSON.stringify(expected), done);
    });

    it('should respond with 404 status when the requested resource does not exist', function(done) {
        var app = express(),
            expected = {
                message: 'Not Found'
            };

        app.get('/api/test', function(req, res) {
            api.notFound(req, res);
        });

        request(app)
            .get('/api/test')
            .expect('Content-Type', /json/)
            .expect(404)
            .expect(JSON.stringify(expected), done);
    });

    it('should respond with 405 status when an unsupported API action is invoked', function(done) {
        var app = express(),
            expected = {
                message: 'Unsupported Action'
            };

        app.get('*', function(req, res) {
            api.unsupportedAction(req, res);
        });

        request(app)
            .get('/api/nonexistent')
            .expect('Content-Type', /json/)
            .expect(405)
            .expect(JSON.stringify(expected), done);
    });

    it('should respond with 422 status when validation fails', function(done) {
        var app = express(),
            expected = {
                message: 'Validation Failed',
                errors: ['Username is already taken.']
            };
        app.use(bodyParser.json());
        app.use(bodyParser.urlencoded({
            extended: true
        }));

        app.post('/api/test', function(req, res) {
            if (req.body.username === 'hercules')
                api.invalid(req, res, 'Username is already taken.');
        });

        request(app)
            .post('/api/test')
            .send({
                username: 'hercules'
            })
            .expect('Content-Type', /json/)
            .expect(422)
            .expect(JSON.stringify(expected), done);
    });

    it('should respond with 500 status for internal server errors', function(done) {
        var app = express(),
            expected = {
                message: 'Internal Server Error',
                error: {
                    message: 'Database error',
                    stacktrace: '...'
                }
            };

        app.get('/api/test', function(req, res) {
            var err = new Error('Database error');
            err.stack = '...';
            api.serverError(req, res, err);
        });

        request(app)
            .get('/api/test')
            .expect('Content-Type', /json/)
            .expect(500)
            .expect(JSON.stringify(expected), done);
    });
});


lib/index.js
'use strict';

var _hasOwnProperty = Object.prototype.hasOwnProperty;

var Status = {
    OK: 200,
    BAD_REQUEST: 400,
    UNAUTHORIZED: 401,
    FORBIDDEN: 403,
    NOT_FOUND: 404,
    UNSUPPORTED_ACTION: 405,
    VALIDATION_FAILED: 422,
    SERVER_ERROR: 500
};

function statusMessage(status) {
    switch (status) {
        case Status.BAD_REQUEST:
            return 'Bad Request';
        case Status.UNAUTHORIZED:
            return 'Unauthorized';
        case Status.FORBIDDEN:
            return 'Forbidden';
        case Status.NOT_FOUND:
            return 'Not Found';
        case Status.UNSUPPORTED_ACTION:
            return 'Unsupported Action';
        case Status.VALIDATION_FAILED:
            return 'Validation Failed';
        case Status.SERVER_ERROR:
            return 'Internal Server Error';
    }
}

function jsonResponse(res, body, options) {
    options = options || {};
    options.status = options.status || Status.OK;
    res.status(options.status).json(body || null);
}

var api = {
    ok: function(req, res, data) {
        jsonResponse(res, data, {
            status: Status.OK
        });
    },

    badRequest: function(req, res, errors) {
        errors = Array.isArray(errors) ? errors : [errors];

        var body = {
            message: statusMessage(Status.BAD_REQUEST),
            errors: errors
        };

        jsonResponse(res, body, {
            status: Status.BAD_REQUEST
        });
    },

    unauthorized: function(req, res) {
        var body = {
            message: statusMessage(Status.UNAUTHORIZED)
        };

        jsonResponse(res, body, {
            status: Status.UNAUTHORIZED
        });
    },

    forbidden: function(req, res) {
        var body = {
            message: statusMessage(Status.FORBIDDEN)
        };

        jsonResponse(res, body, {
            status: Status.FORBIDDEN
        });
    },
    notFound: function(req, res) {
        var body = {
            message: statusMessage(Status.NOT_FOUND)
        };

        jsonResponse(res, body, {
            status: Status.NOT_FOUND
        });
    },

    unsupportedAction: function(req, res) {
        var body = {
            message: statusMessage(Status.UNSUPPORTED_ACTION)
        };

        jsonResponse(res, body, {
            status: Status.UNSUPPORTED_ACTION
        });
    },

    invalid: function(req, res, errors) {
        errors = Array.isArray(errors) ? errors : [errors];

        var body = {
            message: statusMessage(Status.VALIDATION_FAILED),
            errors: errors
        };

        jsonResponse(res, body, {
            status: Status.VALIDATION_FAILED
        });
    },

    serverError: function(req, res, error) {
        if (error instanceof Error) {
            error = {
                message: error.message,
                stacktrace: error.stack
            };
        }
        var body = {
            message: statusMessage(Status.SERVER_ERROR),
            error: error
        };

        jsonResponse(res, body, {
            status: Status.SERVER_ERROR
        });
    },

    requireParams: function(req, res, params, next) {
        var missing = [];

        params = Array.isArray(params) ? params : [params];

        params.forEach(function(param) {
            if (!(req.body && _hasOwnProperty.call(req.body, param)) && !(req.params && _hasOwnProperty.call(req.params, param)) && !_hasOwnProperty.call(req.query, param)) {
                missing.push('Missing required parameter: ' + param);
            }
        });

        if (missing.length) {
            api.badRequest(req, res, missing);
        } else {
            next();
        }
    },

    requireHeaders: function(req, res, headers, next) {
        var missing = [];

        headers = Array.isArray(headers) ? headers : [headers];

        headers.forEach(function(header) {
            if (!(req.headers && _hasOwnProperty.call(req.headers, header))) {
                missing.push('Missing required header parameter: ' + header);
            }
        });

        if (missing.length) {
            api.badRequest(req, res, missing);
        } else {
            next();
        }
    }
};

module.exports = api;
